source/module_io/read_cube.cpp

```diff
@@ -147,43 +140,51 @@ void ModuleIO::trilinear_interpolate(
     delete[] read_rho;
 }
 
-ModuleIO::CubeInfo ModuleIO::read_cube(const std::string& file)
+bool ModuleIO::read_cube(const std::string& file,
+    std::vector<std::string>& comment,
+    int& natom,
+    std::vector<double>& cell_pos,
+    std::vector<int>& nvoxel,
+    std::vector<std::vector<double>>& axis_vecs,
+    std::vector<int>& atom_type,
+    std::vector<double>& atom_charge,
+    std::vector<std::vector<double>>& atom_pos,
+    std::vector<double>& data)
 {
     std::ifstream ifs(file);
 
-    if (!ifs) { return ModuleIO::CubeInfo({}, 0, {}, {}, {}, {}, {}, {}, {}, false); }
+    if (!ifs) { return false; }
 
-    std::vector<std::string> comment(2);
+    comment.resize(2);
     for (auto& c : comment) { std::getline(ifs, c); }
 
-    int natom;
     ifs >> natom;
-    std::vector<double> cell_pos(3);
+    cell_pos.resize(3);
     for (auto& cp : cell_pos) { ifs >> cp; }
 
-    std::vector<int> nvoxel(3);
-    std::vector<std::vector<double>> axis_vecs(3);
+    nvoxel.resize(3);
+    axis_vecs.resize(0);
     for (int i = 0;i < 3;++i)
     {
         std::vector<double> vec(3);
         ifs >> nvoxel[i] >> vec[0] >> vec[1] >> vec[2];
         axis_vecs.push_back(vec);
     }
 
-    std::vector<int> itype(natom);
-    std::vector<double> charge(natom);
-    std::vector<std::vector<double>> atom_pos(natom);
+    atom_type.resize(natom);
+    atom_charge.resize(natom);
+    atom_pos.resize(0);
```

Comment:
you already have `natom`, so you can resize it to natom first, because the `push_back` is always much slower than allocate-assignment way, because `push_back` may trigger the memory release and reallocation for many times.

Reply:
or if you dont like to do like this, you can add an assertion that
```c++
assert(natom == atom_pos.size())
```
then your function will be safer.